<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Temi" />

<meta name="date" content="2022-05-13" />

<title>Simulating discrete and continuous populations</title>

<script src="site_libs/header-attrs-2.14/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>

<link rel="icon" href="https://github.com/workflowr/workflowr-assets/raw/main/img/reproducible.png">
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>



<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>









<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.tab('show');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html"></a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="/">home</a>
</li>
<li>
  <a href="/projects/">projects</a>
</li>
<li>
  <a href="/blog/">thoughts</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">Simulating discrete and continuous
populations</h1>
<h4 class="author">Temi</h4>
<h4 class="date">2022-05-13</h4>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span>
workflowr <span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span
class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> </a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2022-05-13
</p>
<p>
<strong>Checks:</strong> <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 5
<span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> 2
</p>
<p>
<strong>Knit directory:</strong> <code>isiro/</code> <span
class="glyphicon glyphicon-question-sign" aria-hidden="true"
title="This is the local directory in which the code in this file was executed.">
</span>
</p>
<p>
This reproducible <a href="https://rmarkdown.rstudio.com">R Markdown</a>
analysis was created with <a
  href="https://github.com/workflowr/workflowr">workflowr</a> (version
1.7.0). The <em>Checks</em> tab describes the reproducibility checks
that were applied when the results were created. The <em>Past
versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguncommittedchanges">
<span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> <strong>R Markdown file:</strong> uncommitted
changes </a>
</p>
</div>
<div id="strongRMarkdownfilestronguncommittedchanges"
class="panel-collapse collapse">
<div class="panel-body">
<p>The R Markdown file has unstaged changes. To know which version of
the R Markdown file created these results, you’ll want to first commit
it to the Git repo. If you’re still working on the analysis, you can
ignore this warning. When you’re finished, you can run
<code>wflow_publish</code> to commit the R Markdown file and build the
HTML.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the
global environment can affect the analysis in your R Markdown file in
unknown ways. For reproduciblity it’s best to always run the code in an
empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed20220512code">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Seed:</strong>
<code>set.seed(20220512)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed20220512code"
class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(20220512)</code> was run prior to running
the code in the R Markdown file. Setting a seed ensures that any results
that rely on randomness, e.g. subsampling or permutations, are
reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Session information:</strong>
recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package
versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongdetected">
<span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> <strong>Cache:</strong> detected </a>
</p>
</div>
<div id="strongCachestrongdetected" class="panel-collapse collapse">
<div class="panel-body">
The following chunks had caches available:
<ul>
<li>
session-info-chunk-inserted-by-workflowr
</li>
<li>
unnamed-chunk-1
</li>
<li>
unnamed-chunk-2
</li>
<li>
unnamed-chunk-3
</li>
<li>
unnamed-chunk-4
</li>
<li>
unnamed-chunk-5
</li>
<li>
unnamed-chunk-6
</li>
<li>
unnamed-chunk-7
</li>
</ul>
<p>To ensure reproducibility of the results, delete the cache directory
<code>simulating_populations_cache</code> and re-run the analysis. To
have workflowr automatically delete the cache directory prior to
building the file, set <code>delete_cache = TRUE</code> when running
<code>wflow_build()</code> or <code>wflow_publish()</code>.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongrelative">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>File paths:</strong> relative </a>
</p>
</div>
<div id="strongFilepathsstrongrelative" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Using relative paths to the files within your workflowr
project makes it easier to run your code on other machines.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgithubcomtemipeteisirotree84bb38ada0619b031c5c3ef057fdd7fe6674772ctargetblank84bb38aa">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Repository version:</strong>
<a href="https://github.com/temipete/isiro/tree/84bb38ada0619b031c5c3ef057fdd7fe6674772c" target="_blank">84bb38a</a>
</a>
</p>
</div>
<div
id="strongRepositoryversionstrongahrefhttpsgithubcomtemipeteisirotree84bb38ada0619b031c5c3ef057fdd7fe6674772ctargetblank84bb38aa"
class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development
and connecting the code version to the results is critical for
reproducibility.
</p>
<p>
The results in this page were generated with repository version
<a href="https://github.com/temipete/isiro/tree/84bb38ada0619b031c5c3ef057fdd7fe6674772c" target="_blank">84bb38a</a>.
See the <em>Past versions</em> tab to see a history of the changes made
to the R Markdown and HTML files.
</p>
<p>
Note that you need to be careful to ensure that all relevant files for
the analysis have been committed to Git prior to generating the results
(you can use <code>wflow_publish</code> or
<code>wflow_git_commit</code>). workflowr only checks the R Markdown
file, but you know if there are other scripts or data files that it
depends on. Below is the status of the Git repository when the results
were generated:
</p>
<pre><code>
Ignored files:
    Ignored:    .Rhistory
    Ignored:    analysis/simulating_populations_cache/

Unstaged changes:
    Modified:   analysis/simulating_populations.Rmd

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not
included in this status report because it is ok for generated content to
have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">

<p>
These are the previous versions of the repository in which changes were
made to the R Markdown
(<code>analysis/simulating_populations.Rmd</code>) and HTML
(<code>docs/simulating_populations.html</code>) files. If you’ve
configured a remote Git repository (see <code>?wflow_git_remote</code>),
click on the hyperlinks in the table below to view the files as they
were in that past version.
</p>
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
File
</th>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
<th>
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/temipete/isiro/84bb38ada0619b031c5c3ef057fdd7fe6674772c/docs/simulating_populations.html" target="_blank">84bb38a</a>
</td>
<td>
Temi
</td>
<td>
2022-05-13
</td>
<td>
Updated website
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/temipete/isiro/05f359b65aefe45d79d7eb64c62af43bd7e918e1/docs/simulating_populations.html" target="_blank">05f359b</a>
</td>
<td>
Temi
</td>
<td>
2022-05-13
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/temipete/isiro/4bc77f948d264460ae3aec941f482ff5f7233e6f/docs/simulating_populations.html" target="_blank">4bc77f9</a>
</td>
<td>
Temi
</td>
<td>
2022-05-13
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/temipete/isiro/e7368915c2c44318eda54a3ddb5bdcdfbf10eb94/docs/simulating_populations.html" target="_blank">e736891</a>
</td>
<td>
Temi
</td>
<td>
2022-05-13
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/temipete/isiro/blob/438da91febd13664a1515bf56565fbd33de0e514/analysis/simulating_populations.Rmd" target="_blank">438da91</a>
</td>
<td>
Temi
</td>
<td>
2022-05-13
</td>
<td>
wflow_publish("./analysis/*.Rmd")
</td>
</tr>
</tbody>
</table>
</div>
<hr>
</div>
</div>
</div>
<div id="im-still-working-on-this" class="section level3">
<h3>I’m still working on this</h3>
<pre class="r"><code>library(gtools)</code></pre>
<pre class="r"><code>source(&#39;https://raw.githubusercontent.com/TemiPete/miscellaneous_code/main/R/misc_functions.R&#39;)</code></pre>
</div>
<div id="introduction" class="section level2">
<h2>Introduction</h2>
<p>Oftentimes, we want to simulate populations in order to do some
analysis. In this demo, I go over a, perhaps, crude way to mimic
discrete and continuous populations. Before simulating, we need to know
what a population is. In genetics, a population is a grouping of
individuals based on some similarity in allele frequencies between them.
Of course, there are many other definitions of what a population is. In
this demo, I will use the definition based on allele frequencies.</p>
<p>One way to generate allele frequencies that mimic that seen in
typical populations is to use the <span
class="math inline">\(F_{ST}\)</span>, AKA the fixation index, AKA
F-statistic. <span class="math inline">\(F_{ST}\)</span> measures
population differentiation as a result of genetic structure. Values
range between 0 and 1, and higher values mean that two populations are
highly similar. There are many definitions of <span
class="math inline">\(F_{ST}\)</span>, depending on what you are
discussing - and I am still learning about most of them. In this case,
however, I am primarily concerned with allele frequency-related
definitions.</p>
</div>
<div id="code" class="section level2">
<h2>Code</h2>
<p>To simulate populations based on the definition of <span
class="math inline">\(F_{ST}\)</span>, one thing we can do is to
simulate an ancestral population’s allele frequencies. This is similar
to having a founder population from which every other population
descends. Afterwards, we can simulate different populations from this
ancestral population. This follows from the <a
href="https://en.wikipedia.org/wiki/Balding%E2%80%93Nichols_model">Balding-Nichols
model</a>.</p>
<p><span class="math inline">\(Beta(\frac{1-F}{F}p,
\frac{1-F}{F}(1-p))\)</span></p>
<p>Assume that we have 500 alleles/loci, and 4 populations, and we
intend to simulate 1000 individuals from each of these populations. We
can define our <span class="math inline">\(F_{ST}\)</span> to be, say,
0.09 between populations 1 and 2, 0.19 between populations 1 and 2, and
population 3, and 0.4 between all these populations, and population 4.
Therefore, populations 1 and 2 will be closely related, and different
from population 3, and all of them will be different from population
4.</p>
<pre class="r"><code>M &lt;- 500 # number of alleles or SNPs
n_pops &lt;- 4 # number of populations
N &lt;- 1000 # 20 individuals in each n_pops
f_pop12 &lt;- 0.09
f_pop3 &lt;- 0.19 # a third, distant, unrelated population
f_pop4 &lt;- 0.4 # a fourth population</code></pre>
<p>When we make plots of the allele frequencies of these populations, we
expect high correlations between populations 1 and 2, and not-so-high
correlations when we compare with the other populations 3 and 4.</p>
<p>We can generate some random minor allele frequencies (MAFs) to be the
ancestral allele frequencies, and generate independent draws from the
distribution of the ancestral allele frequencies, based on the equation
above.</p>
<pre class="r"><code>ancestral_allele_freqs &lt;- runif(M, 0.01, 0.5)
sh1 &lt;- ((1-f_pop12)/f_pop12)*ancestral_allele_freqs
sh2 &lt;- ((1-f_pop12)/(f_pop12))*(1-ancestral_allele_freqs)
pop1_allele_freqs &lt;- rbeta(M, shape1 = sh1, shape2 = sh2)
pop2_allele_freqs &lt;- rbeta(M, shape1 = sh1, shape2 = sh2)
pop3_allele_freqs &lt;- rbeta(M, shape1 = ((1-f_pop3)/f_pop3)*ancestral_allele_freqs, shape2=((1-f_pop3)/(f_pop3))*(1-ancestral_allele_freqs))
pop4_allele_freqs &lt;- rbeta(M, shape1 = ((1-f_pop4)/f_pop4)*ancestral_allele_freqs, shape2=((1-f_pop4)/(f_pop4))*(1-ancestral_allele_freqs))

all_allele_freqs &lt;- list(pop1_allele_freqs, pop2_allele_freqs, pop3_allele_freqs, pop4_allele_freqs)
names(all_allele_freqs) &lt;- paste0(&#39;population_&#39;, 1:4)</code></pre>
<pre class="r"><code>layout(matrix(c(1,2,3,4,5,6), nrow = 2, ncol = 3, byrow = T))
names_pop &lt;- names(all_allele_freqs)
for(i in 1:4){
    for(j in 1:4){
        if(i == j){ # we don&#39;t want to plot pop1 vs pop1 e.t.c.
            next
        } else if (i &lt; j) {
            cor_ &lt;- round(cor(all_allele_freqs[[i]], all_allele_freqs[[j]]), 2)
            plot(all_allele_freqs[[i]], all_allele_freqs[[j]], xlab = names_pop[i], ylab = names_pop[j], frame.plot = F)
            abline(a=0, b=cor_, col=&#39;red&#39;)
            mtext(cor_)
        }
    }
}
mtext(&#39;Scatterplot of allele frequencies for each population&#39;, outer = T)</code></pre>
<p><img src="figure/simulating_populations.Rmd/scatterplot%20allele%20frequencies-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>The plot is approximately what we expect!</p>
</div>
<div id="simulating-genotypes-using-these-allele-frequencies"
class="section level2">
<h2>Simulating genotypes using these allele frequencies</h2>
<p>Given these allele frequencies, we can simulate genotypes from each
of these populations by sampling from a binomial distribution.</p>
<p><span class="math inline">\(G_{k_{n}} \sim Binom(n,
p_{k})\)</span></p>
<p>where <span class="math inline">\(n\)</span> is the number of
success, and <span class="math inline">\(p\)</span> is the probability
of a success. Here, <span class="math inline">\(n = 2\)</span> because
we want 2 alleles per loci, and <span class="math inline">\(p\)</span>
is the allele frequency at that loci.</p>
<pre class="r"><code>genotypes &lt;- lapply(seq_along(all_allele_freqs), function(i){
    matrix(data=rbinom(n=N*M, size=2, prob = all_allele_freqs[[i]]), nrow = N, ncol = M, byrow = T)
})

genotypes &lt;- do.call(rbind, genotypes)
shuffle &lt;- sample(1:nrow(genotypes)) # I will shuffle the data
genotypes &lt;- genotypes[shuffle, ]
pops &lt;- rep(names(all_allele_freqs), each=N)
pops &lt;- pops[shuffle] # shuffle pops too, since it tells me what populations the genotypes come from

cat(&#39;Here is what the genotypes look like\n&#39;)</code></pre>
<pre><code>Here is what the genotypes look like</code></pre>
<pre class="r"><code>genotypes[1:5, 1:5]</code></pre>
<pre><code>     [,1] [,2] [,3] [,4] [,5]
[1,]    0    0    0    0    1
[2,]    0    0    0    0    2
[3,]    0    0    0    1    2
[4,]    0    0    0    0    0
[5,]    0    0    0    0    0</code></pre>
<pre class="r"><code>cat(&#39;And here is what the population designation looks like \n&#39;)</code></pre>
<pre><code>And here is what the population designation looks like </code></pre>
<pre class="r"><code>pops[1:5]</code></pre>
<pre><code>[1] &quot;population_1&quot; &quot;population_4&quot; &quot;population_2&quot; &quot;population_1&quot; &quot;population_4&quot;</code></pre>
<p>Next, we can calculate the principal components (PCs) of this data,
and plot PC1 vs PC2, to see where the most variability lies in the
data</p>
<pre class="r"><code># first change the row names of the genotype data to reflect the populations
rownames(genotypes) &lt;- pops
# I need to make sure that there are some variations in the SNPS - if there aren&#39;t any, I should remove them
pca_genotypes &lt;- genotypes[ , which(apply(genotypes, 2, var) != 0)]
# PCA
pca_dt &lt;- prcomp(pca_genotypes, scale.=T, center=T)
pca_dt$x[1:5, 1:5]</code></pre>
<pre><code>                    PC1        PC2         PC3         PC4         PC5
population_1   3.365441 -3.5048602  6.53058793  3.64826678 -1.55650504
population_4 -12.673512  0.4779913 -0.01039136 -0.82388450 -0.02560419
population_2   3.934678 -3.3804340 -5.75174919  0.04647499 -0.63133432
population_1   3.557170 -5.6361956  5.75008629  0.91973143 -1.34829976
population_4 -13.115938 -0.2937992 -0.22428779  0.17730445 -1.79836083</code></pre>
<pre class="r"><code># colors
mycolors &lt;- colorRampPalette(RColorBrewer::brewer.pal(8, &quot;Dark2&quot;))(n_pops)
#mycolors &lt;- c(&#39;red&#39;, &#39;orange&#39;, &#39;green&#39;, &#39;blue&#39;)
color_coding &lt;- factor(rownames(pca_dt$x), labels = mycolors)
plot(pca_dt$x[, &#39;PC1&#39;], pca_dt$x[, &#39;PC2&#39;], bg=as.character(color_coding), xlab = &#39;PC1&#39;, ylab = &#39;PC2&#39;, frame.plot = F, pch=21, main=&#39;PC1 vs PC2 of 4 discrete populations&#39;)
legend(&#39;topleft&#39;, legend=names_pop, pch=c(21, 21, 21, 21), pt.bg=mycolors, bty = &#39;n&#39;, xpd=NA)</code></pre>
<p><img src="figure/simulating_populations.Rmd/unnamed-chunk-6-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>We can look at populations 1 and 2 only, and see if there is any
difference between them.</p>
<pre class="r"><code>pca_12 &lt;- pca_dt$x[rownames(pca_dt$x) %in% c(&#39;population_1&#39;, &#39;population_2&#39;), ]
plot(pca_12[, &#39;PC1&#39;], pca_12[, &#39;PC2&#39;], col=rownames(pca_12) |&gt; as.factor(), bg=rownames(pca_12) |&gt; as.factor(), xlab = &#39;PC1&#39;, ylab = &#39;PC2&#39;, frame.plot = F, pch=19)</code></pre>
<p><img src="figure/simulating_populations.Rmd/unnamed-chunk-7-1.png" width="672" style="display: block; margin: auto;" /></p>
</div>
<div id="simulating-continuous-populations" class="section level2">
<h2>Simulating continuous populations</h2>
<p>What I have done so far, is to simulate a discrete population of
individuals. Think of it as people living in Mars, Earth, and Jupiter,
and who have never intermingled since time immemorial. Population 3 is
faraway from population 4. However, true human populations are not this
way. How can we simulate a truly continuous population?</p>
<p>First, notice that I am only simulating allele frequencies.
Therefore, I can put a prior on my earlier-defined beta distribution to
model the probability that an individual is from population <span
class="math inline">\(m\)</span>. In other words, I am trying to
estimate what percentage of an individual’s alleles might have come from
a certain population <span class="math inline">\(m\)</span>. I will call
this random variable, <span class="math inline">\(Q\)</span>.</p>
<p><span class="math inline">\(Q \sim Dirichlet(\alpha)\)</span></p>
<p>A trick I can use is to take an individual, ask how much of their
genome is shared among each population I have, select that proportion of
allele frequencies, and simulate their genotypes based on the selection
for each population. To do this, I need to know how to distribute the
<span class="math inline">\(\alpha\)</span> parameter used in the
Dirichlet distribution. A trick I can use is to use the allele frequency
distribution in the entire population. I will call this distribution the
population grade distribution.</p>
<p>Earlier, I simulated some allele frequencies
<code>all_allele_freqs</code></p>
<pre class="r"><code>ss &lt;- sapply(all_allele_freqs, sum)
population_grade &lt;- ss/sum(ss)
#population_grade &lt;- rand_vect_cont(4, 1)
population_grade</code></pre>
<pre><code>population_1 population_2 population_3 population_4 
   0.2550745    0.2484377    0.2444566    0.2520312 </code></pre>
<p>Then I can use this as a parameter to the Dirichlet distribution to
simulate the percentage of an individual’s genotypes that is shared
across all 4 populations.</p>
<pre class="r"><code>q_matrix &lt;- gtools::rdirichlet(N * n_pops, alpha = population_grade) # I want to simulate, N individuals per population</code></pre>
<p>Here is an illustration. Assuming that I am looking at the first
individual</p>
<pre class="r"><code># assuming one individual
id1 &lt;- q_matrix[1, ]
q_counts &lt;- floor(id1 * M) # number of loci shared
q_counts</code></pre>
<pre><code>[1]   0 273 101 124</code></pre>
<p>The assumption here is that this individual shares 0 loci with
population 1, 273 with population 2, 101 with population 3, and 124 loci
with population 4, and 3 independent loci not share with any population.
The sum is 498 (because I rounded down), but I don’t want this to be
this case. Therefore, I will share the remaining loci with the highest
number already available, so that the sum is 500.</p>
<pre class="r"><code>remainder_ &lt;- M - sum(q_counts)
q_counts[which.max(q_counts)] &lt;- q_counts[which.max(q_counts)] + remainder_
q_counts</code></pre>
<pre><code>[1]   0 275 101 124</code></pre>
<p>Now, this sums to 500.</p>
<p>Putting this into a reusable function…</p>
<pre class="r"><code>q_counts_fxn &lt;- function(q_matrix, M){
    apply(q_matrix, 1, function(each_row){
        temp &lt;- floor(each_row * M)
        remainder_ &lt;- M - sum(temp)
        temp[which.max(temp)] &lt;- temp[which.max(temp)] + remainder_
        temp
    }) |&gt; t()
}

q_counts &lt;- q_counts_fxn(q_matrix, M)
q_counts[1:5, ] # for the first 5 individuals</code></pre>
<pre><code>     [,1] [,2] [,3] [,4]
[1,]    0  275  101  124
[2,]   28  472    0    0
[3,]  373    0   70   57
[4,]  495    0    3    2
[5,]  130  197    1  172</code></pre>
<pre class="r"><code>genotypes &lt;- matrix(NA, nrow = N*n_pops, ncol = M) # remember, N individuals per population

for(q in 1:nrow(q_counts)){
    pop_down &lt;- 1
    m_loci &lt;- 1:M
    i_loci &lt;- list() # this list contains the actual loci an individual has shared among the other populations
    while(pop_down &lt;= ncol(q_counts)){
        # sample pop 1 loci
        pop_loci &lt;- sample(m_loci, size=q_counts[q, pop_down], replace = F)
        if(length(pop_loci) == 0){
            m_loci &lt;- m_loci[m_loci != 0]
            i_loci[[pop_down]] &lt;- 0
        } else {
            m_loci &lt;- m_loci[!m_loci %in% pop_loci]
            i_loci[[pop_down]] &lt;- pop_loci
        }
        pop_down &lt;- pop_down + 1
    }
    
    g_loci &lt;- lapply(seq_along(i_loci), function(i){
        temp &lt;- all_allele_freqs[[i]][i_loci[[i]]]
        rbinom(n = length(temp), size = 2, prob = temp)
    })

    gi &lt;- vector(mode=&#39;numeric&#39;, length = M)
    for(i in 1:n_pops){
        gi[i_loci[[i]]] &lt;- g_loci[[i]]
    }
    
    genotypes[q, ] &lt;- gi
    
}</code></pre>
<pre class="r"><code>pop &lt;- paste0(&#39;population_&#39;, apply(q_counts, 1, which.max))
# first change the row names of the genotype data to reflect the populations
rownames(genotypes) &lt;- pop
# I need to make sure that there are some variations in the SNPS - if there aren&#39;t any, I should remove them
pca_genotypes &lt;- genotypes[ , which(apply(genotypes, 2, var) != 0)]
# PCA
pca_dt &lt;- prcomp(pca_genotypes, scale.=T, center=T)
pca_dt$x[1:5, 1:5]</code></pre>
<pre><code>                    PC1        PC2       PC3        PC4         PC5
population_2  0.6843511  0.7775393 -1.986994 -0.5733985  1.35030907
population_2  5.2284319 -4.4591500 -5.521154  1.0043617 -0.98073989
population_1  1.2348129 -0.8818382  4.936476  0.7053044 -0.35913221
population_1  3.5599093 -3.7287330  5.498686 -4.3138313  0.46221479
population_2 -1.7057116 -2.4540082 -1.934744  0.7724849 -0.06324158</code></pre>
<pre class="r"><code>mycolors &lt;- colorRampPalette(RColorBrewer::brewer.pal(8, &quot;Dark2&quot;))(n_pops)
color_coding &lt;- factor(rownames(pca_dt$x), labels = mycolors)
plot(pca_dt$x[, &#39;PC1&#39;], pca_dt$x[, &#39;PC2&#39;], bg=color_coding |&gt; as.character(), xlab = &#39;PC1&#39;, ylab = &#39;PC2&#39;, frame.plot = F, pch=21, main=&#39;PC1 vs PC2 of 4 continuous populations&#39;)
legend(&#39;topleft&#39;, legend=names_pop, pch=c(21, 21, 21, 21), pt.bg=mycolors, bty = &#39;n&#39;, xpd = NA)</code></pre>
<p><img src="figure/simulating_populations.Rmd/plotting%20continuous%20population-1.png" width="672" style="display: block; margin: auto;" /></p>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span>
Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 4.1.3 (2022-03-10)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Pop!_OS 22.04 LTS

Matrix products: default
BLAS:   /usr/lib/x86_64-linux-gnu/blas/libblas.so.3.10.0
LAPACK: /usr/lib/x86_64-linux-gnu/lapack/liblapack.so.3.10.0

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] gtools_3.9.2

loaded via a namespace (and not attached):
 [1] Rcpp_1.0.8.3       RColorBrewer_1.1-3 pillar_1.7.0       compiler_4.1.3    
 [5] bslib_0.3.1        later_1.3.0        jquerylib_0.1.4    git2r_0.30.1      
 [9] highr_0.9          workflowr_1.7.0    tools_4.1.3        digest_0.6.29     
[13] jsonlite_1.8.0     evaluate_0.15      lifecycle_1.0.1    tibble_3.1.7      
[17] pkgconfig_2.0.3    rlang_1.0.2        cli_3.3.0          rstudioapi_0.13   
[21] yaml_2.3.5         xfun_0.31          fastmap_1.1.0      stringr_1.4.0     
[25] knitr_1.39         fs_1.5.2           vctrs_0.4.1        sass_0.4.1        
[29] rprojroot_2.0.3    glue_1.6.2         R6_2.5.1           fansi_1.0.3       
[33] rmarkdown_2.14     magrittr_2.0.3     whisker_0.4        promises_1.2.0.1  
[37] ellipsis_0.3.2     htmltools_0.5.2    httpuv_1.6.5       utf8_1.2.2        
[41] stringi_1.7.6      crayon_1.5.1      </code></pre>
</div>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
https://docs.mathjax.org/en/latest/web/configuration.html. This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>




</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
